java/ql/src/semmle/code/java/dataflow/internal/DataFlowUtil.qll

```diff
@@ -152,6 +150,14 @@ predicate simpleLocalFlowStep(Node node1, Node node2) {
   )
   or
   FlowSummaryImpl::Private::Steps::summaryLocalStep(node1, node2, true)
+  or
+  // If flow through a method updates a parameter from some input A, and that
+  // parameter also is returned through B, then we'd like a combined flow from A
+  // to B as well. As an example, this simplifies modeling of fluent methods:
+  // for `StringBuilder.append(x)` with a specified value flow from qualifier to
+  // return value and taint flow from argument 0 to the qualifier, then this
+  // allows the inferral of taint flow from argument 0 to the return value.
+  node1.(SummaryNode).(PostUpdateNode).getPreUpdateNode().(ParameterNode) = node2
```

Comment:
Won't this allow for "time travelling"?

Reply:
A little, yes. Incorporating flow from aliasing always does, but I don't think it'll result in spurious flow outside some very dedicated crafted examples. In particular, the pre-update node is not in an argument position, which would be the case most likely to exhibit time travel. Normally, I'd expect post-update nodes to have mostly the same set of local successors through simple flow steps as their pre-update node, but in this case it mostly amounts to the same thing as simply stepping directly to the pre-update node. Additionally, multiple flow models inside a `SummarizedCallable` don't really have associated control-flow or relative ordering, so I think we're staying within the precision-level that the `SummarizedCallable` models allow.
Are you able to think of any plausible examples with weird time-travel effects?

Reply:
Ah, I misread `ParameterNode` as `ArgumentNode`. I wonder if the line above should be added to the shared library?